be/src/storage/compaction_utils.cpp

```diff
@@ -68,15 +68,12 @@ Status CompactionUtils::construct_output_rowset_writer(Tablet* tablet, uint32_t
 
 uint32_t CompactionUtils::get_segment_max_rows(int64_t max_segment_file_size, int64_t input_row_num,
                                                int64_t input_size) {
-    // max segment rows
-    int64_t max_segment_rows = max_segment_file_size * input_row_num / (input_size + 1);
-    // default value in RowsetWriterContext is INT32_MAX.
-    // segment file use uint32 to represent row number,
-    // and use INT32_MAX to avoid overflow issue when casting from uint32_t to int.
+    // The max of config::max_segments_file_size is INT_MAX
+    // The max of input_row_num is UINT32_MAX
+    // Using division before multiplication can avoid the overflow
+    int64_t max_segment_rows = max_segment_file_size * (input_row_num / input_size + 1);
```

Comment:
Why not use double to compute this number?
```suggestion
    int64_t max_segment_rows = 1.0 * max_segment_file_size * input_row_num / (input_size + 1);
```